cpp/src/arrow/compute/exec/union_node.cc

```diff
@@ -0,0 +1,131 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <mutex>
+
+#include "arrow/api.h"
+#include "arrow/compute/api.h"
+#include "arrow/compute/exec/exec_plan.h"
+#include "arrow/compute/exec/options.h"
+#include "arrow/compute/exec/util.h"
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/future.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+
+namespace compute {
+
+struct UnionNode : ExecNode {
+  UnionNode(ExecNode* lhs_input, ExecNode* rhs_input, ExecContext* ctx)
+      : ExecNode(lhs_input->plan(), {lhs_input, rhs_input},
+                 {"left_input_union", "right_input_union"},
+                 /*output_schema=*/lhs_input->output_schema(),
+                 /*num_outputs=*/1),
+        ctx_(ctx) {}
+
+  const char* kind_name() override { return "UnionNode"; }
+
+  inline bool IsLeftInput(ExecNode* input) { return input == inputs_[0]; }
+
+  void InputReceived(ExecNode* input, int seq, ExecBatch batch) override {
+    ARROW_DCHECK(input == inputs_[0] || input == inputs_[1]);
+
+    if (finished_.is_finished()) {
+      return;
+    }
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
I think it would be better to use AtomicCounter here 
https://github.com/apache/arrow/blob/d943b8bf22149775c84b5afbda5c34cb9a19d93c/cpp/src/arrow/compute/exec/util.h#L194